### DISCUSSION TITLE
Is it possible to make `superRefine` check nested fields?

### DISCUSSION THREAD

**Discussion by sneko** [colinhacks/zod#1947] [Q&A] [answered]
Hi,

My checks work but I feel it misses some information. I bind 2 booleans, if `main` is not true, `additional` must be `null`. I can do it by using `refine` or `superRefine` on the parent object, like so:
```ts

const ExampleSchema = z
  .object({
    hello: z.string().min(1),
    main: z.boolean(),
    additional: z.boolean().nullable(),
  }).superRefine((data, ctx) => {
    if (data && data.main === false && data.additional !== null) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: `XXXXXXXXX`,
      });
    }
  });
```

The issue is if I parse:
```
{
hello:'',
main: false,
additional: true,
}
```

It will return `XXXXXXXXX` but it does not set the error about `hello` being required.

Do you know a way to do so? Since it's for a form, I don't want a first submit to say 1 error (about booleans), and after fixing it there is the `hello` error.

It seems obvious doing the check on the `additional` property directly would work, but when using `refine/superRefine` on it in `data` and `context` I have no way to retrieve the value of `main`.

Thank you,

**Comment by JacobWeisenburger** [comment, marked as answer]
Here's the full example, but I'm not sure what is the problem. Everything seems to be working correctly to me. Can you explain a little more?
```ts
const ExampleSchema = z
    .object({
        hello: z.string().min(1),
        main: z.boolean(),
        additional: z.boolean().nullable(),
    })
    .superRefine((data, ctx) => {
        if (data && data.main === false && data.additional !== null) {
            ctx.addIssue({
                code: z.ZodIssueCode.custom,
                message: `XXXXXXXXX`,
            })
        }
    })

const result = ExampleSchema.safeParse({
    hello: '',
    main: false,
    additional: true,
})
!result.success && console.log(result.error.issues)
// [
//     {
//       code: "too_small",
//       minimum: 1,
//       type: "string",
//       inclusive: true,
//       exact: false,
//       message: "String must contain at least 1 character(s)",
//       path: [ "hello" ]
//     },
//     { code: "custom", message: "XXXXXXXXX", path: [] }
// ]
```

**Reply by sneko** [reply]
@JacobWeisenburger ... just thank you. Was not a good idea to work late haha...

That's my bad I was doing:
```
      CaseSchema.parse({
        ...cases[0],
        citizenId: '',
        alreadyRequestedInThePast: false,
        gotAnswerFromPreviousRequest: true,
      });
```

except that my fixture `cases[0]` was also parsing an object but with a valid "citizenId"... since using `parse()` I didn't see in the logs the difference, thought it was coming from my code above. Big sorry about the waste of time. Thank you for taking time, it made me look at the right place!

**Reply by sneko** [reply]
@JacobWeisenburger just noticed `refine/superRefine` "truncates" the underlying definition since I can not longer access `.shape`.

If I do:
```
ExampleSchema.shape.hello
```

I get: 
```
Property 'shape' does not exist on type ...
```

Should I manage at the end 2 schemas? One to reuse properties elsewhere (with `.shape.xxx`) and another one for parsing (this one would have the `superRefine`)?

**Reply by JacobWeisenburger** [reply]
That is one way to solve it
